### PR TITLE
feat: device modal updates

### LINKED ISSUE #1576: Implement "cancel" action in modals
new trezord (2.0.29?) will allow us to "cancel" actions on display (like address validation, confirmation, etc..)
- [ ] Implement cancel action in modals (trezor-connect)
- [ ] Add new bridge to desktop build

Product:
Decide how to deal with older bridge (it will not work)

### LINKED ISSUE #5237: "Device" modal improvements
**Design:**
Untitled

**Spec:** [Notion](https://www.notion.so/satoshilabs/Device-modal-improvements-c370e7c20fb541a9bdf67b722a973e6f)
- [ ] Remove guide icon from this modal where there is no context and makes not much sense to open guide and start searching.
- [ ] Enable to switch title and image and have the image first and then the title.

### PR DESCRIPTION
**Pull request by dahaca**
closes #5237 
closes #1576

### CODE REVIEW

**Comment by mroz22** [comment]
@trezor/qa pls could you test this with bridge 2.0.27 also?

**Comment by bosomt** [comment]
**QA OK** for 2.0.31 Bridge /that's embedded to Suite binary/ 
**QA NOK** for 2.0.27 Bridge /that's offered for web Suite users/

2.0.27: When you cancel action via modal Suite reports: `Error: device disconnected during action`
2.0.31: When you cancel action via modal Suite reports: `Error: Action cancelled by user`

**Info:**
 - Suite version: desktop 22.6.0 (d1d62b52efdb7aabedef0c30472e57c76765bbb4)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/22.6.0 Chrome/94.0.4606.81 Electron/15.3.0 Safari/537.36
 - OS: MacIntel
 - Screen: 1680x1050
 - Device: model T 2.5.0 regular (revision fd4f8601cf73c5744978f0e8c0aa43d863032a37)
 - Transport: bridge 2.0.31

**Comment by mroz22** [comment]
@bosomt thanks. I shall look into this a bit. Also we might want to reconsider marking this bridge version as outdated with update required. It would solve a bunch of issues. What does @tsusanka  think about this?

File: packages/suite/src/views/firmware/index.tsx [hunk 1]
```diff
@@ -34,15 +34,18 @@ const Wrapper = styled.div<{ isWithTopPadding: boolean }>`
...
         margin-top: 0;
         height: 100%;
+
         > * {
             height: 100%;
         }
     }
-    ${Modal.Content} {
+
+    > ${Modal.Content} {
```
Comment (marekrjpolak) [inline packages/suite/src/views/firmware/index.tsx]: `Content` is not direct child of `Modal`, so this rule is currently unused.

File: packages/components/src/components/prompts/ConfirmOnDevice/index.tsx [hunk 2]
```diff
@@ -111,61 +112,67 @@ const isStepActive = (index: number, activeStep?: number) => {
...
-        </Middle>
-        <Right>
-            <CloseWrapper>
-                {onCancel && (
-                    <Close onClick={onCancel}>
-                        <Icon icon="CROSS" size={23} />
-                    </Close>
+    isConfirmed,
+}: ConfirmOnDeviceProps) => {
+    const hasSteps = steps && activeStep;
```
Comment (marekrjpolak) [inline packages/components/src/components/prompts/ConfirmOnDevice/index.tsx]: Is it safe? It seems that it doesn't happen anywhere, but should `steps === 1` and `activeStep === 0` mean `hasSteps === false`?

File: packages/suite/src/components/suite/images/images.ts [hunk 3]
```diff
@@ -109,63 +55,53 @@ const SVG_IMAGES = {
...
-
-    ${props =>
-        props.theme.IMAGE_FILTER &&
-        css`
-            filter: ${props.theme.IMAGE_FILTER};
-        `}
-`;
-
-const ImageComponent = ({ image, ...props }: Props) => {
-    if (image in PNG_IMAGES) {
```
Comment (marekrjpolak) [inline packages/suite/src/components/suite/images/images.ts:150]: This seems like a perfect spot for typeguards. I'd do something like
```const isPNG = (image: ImageType): image is PngImage => image in PNG_IMAGES;```
and then just `if (isPNG(image)) {...` without `as` assertions.

File: packages/suite/src/components/firmware/CheckSeed.tsx [hunk 4]
```diff
@@ -6,10 +6,8 @@ import { Translation } from '@suite-components';
 import { OnboardingStepBox } from '@onboarding-components';
 import { P } from '@firmware-components';
 
-const CheckboxRow = styled.div`
-    display: flex;
-    flex-direction: row;
-    justify-content: center;
+const StyledChechbox = styled(Checkbox)`
```
Comment (marekrjpolak) [inline packages/suite/src/components/firmware/CheckSeed.tsx]: Chech :)

File: packages/suite/src/components/suite/Modal/DevicePropmptModal.tsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    isConfirmed?: boolean;
+    pillTitle?: string;
+    renderer?: ComponentType<TrezorModalProps>;
+    onAbort?: () => void;
+    className?: string;
+    children?: React.ReactNode;
+    'data-test'?: string;
+}
+
+const DevicePropmptModalRenderer = ({
```
Comment (marekrjpolak) [inline packages/suite/src/components/suite/Modal/DevicePropmptModal.tsx]: Propmpt :upside_down_face:

File: packages/suite/src/components/suite/modals/confirm/ConfirmActionModal.tsx [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+import { H1, variables } from '@trezor/components';
+import { Translation } from '@suite-components/Translation';
+import { DeviceConfirmImage } from '@suite-components/images/DeviceConfirmImage';
+import { TrezorDevice } from '@suite-types';
+import {
+    DevicePropmptModal,
+    DevicePropmptModalProps,
+} from '@suite-components/Modal/DevicePropmptModal';
+
+const StyledDevicePropmptModal = styled(DevicePropmptModal)`
```
Comment (marekrjpolak) [inline packages/suite/src/components/suite/modals/confirm/ConfirmActionModal.tsx]: So many propmpts!

**Review by marekrjpolak — CHANGES_REQUESTED** [review]
image
Should it look like this?

**Comment by tsusanka** [comment]
> 2.0.27: When you cancel action via modal Suite reports: Error: device disconnected during action

Yes, 2.0.27 can't do that.

> Also we might want to reconsider marking this bridge version as outdated with update required. It would solve a bunch of issues. What does @tsusanka think about this?

I would love that, but unfortunately we can not do that at the moment, because the overall Trezor Bridge signing process is broken (https://github.com/trezor/trezord-go/issues/204 and some other issues).

We basically can't release a new standalone Bridge (built-in is just fine, as the app is notarised as a whole) so we are stuck at 27. This sucks tremendously and the plan is basically either to (a) fix it - but we need someone dedicate for Bridge and that is something we were discussing just on Monday together (b) kill standalone Bridge altogether (product discussion needed though!). Before we can decide for either (a) or (b) I would really like to know the result of #4770. Because if that works (b) might be an option (might!).

So tl;dr: I do not think we can roll out this feature to web version of Suite 😞. So I would either postpone this feature or just make it desktop-only (which sounds more reasonable to me). We might theoretically also somehow hack it and make it work on older Bridge versions but I would definitely not spend our energy on that.

File: packages/suite/src/components/suite/images/DeviceConfirmImage.tsx [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+            return 'T_DEVICE_CONFIRM';
+    }
+};
+
+interface DeviceConfirmImageProps extends Omit<ImageProps, 'image'> {
+    device: TrezorDevice;
+}
+
+export const DeviceConfirmImage = ({ device, ...rest }: DeviceConfirmImageProps) => {
+    const majorVersion = device.features ? device.features.major_version : 2;
```
Comment (tomasklim) [inline packages/suite/src/components/suite/images/DeviceConfirmImage.tsx]: It would be nice to use some of the existing methods such as `getDeviceModel` or create a new one to avoid possible inconsistencies in the future.

File: packages/components/src/components/prompts/ConfirmOnDevice/index.tsx [hunk 8]
```diff
@@ -4,25 +4,24 @@ import { DeviceImage } from '../../DeviceImage';
 import { Icon } from '../../Icon';
 import { variables, animations } from '../../../config';
 
-type WrapperProps = Pick<Props, 'animated' | 'animation'>;
-const Wrapper = styled.div<WrapperProps>`
+const Wrapper = styled.div<{ animation?: 'SLIDE_UP' | 'SLIDE_DOWN' }>`
```
Comment (tomasklim) [inline packages/components/src/components/prompts/ConfirmOnDevice/index.tsx]: would be nice to have enum for `'SLIDE_UP' | 'SLIDE_DOWN'`

File: packages/suite/src/components/settings/SettingsMenu.tsx [hunk 9]
```diff
@@ -38,6 +39,49 @@ export const SettingsMenu = () => {
...
+    const appNavItems = useMemo<Array<AppNavigationItem>>(
+        () => [
+            {
+                id: 'settings-index',
+                title: <Translation id="TR_GENERAL" />,
+                position: 'primary',
+                'data-test': '@settings/menu/general',
+                callback: () => {
+                    goto('settings-index', { preserveParams: true });
+                },
```
Comment (tomasklim) [inline packages/suite/src/components/settings/SettingsMenu.tsx]: nit: unnecessary `{ }`

**Review by tomasklim — CHANGES_REQUESTED** [review]
Works well on webusb. 

1. On touch devices, the button should probably not expand because it is shaking otherwise.

https://user-images.githubusercontent.com/33235762/169005897-41725e5c-e1bf-4638-b3e8-d561d4d0369a.mov

2. Actions in onboarding are not cancelable from UI. 

3. 
Not sure, if it is good to show an error modal and error notification when the user cancels action from UI. (Moreover, I do not think it is good to allow cancelling backup as user has to reset the device if he does so)
https://user-images.githubusercontent.com/33235762/169007363-bf7fd6d2-0ee5-4b17-ae65-3a4821243c21.mov

**Comment by mroz22** [comment]
> So tl;dr: I do not think we can roll out this feature to web version of Suite . So I would either postpone this feature or just make it desktop-only (which sounds more reasonable to me). We might theoretically also somehow hack it and make it work on older Bridge versions but I would definitely not spend our energy on that.

1. It is not only web. Might affect desktop as well. Scenario: you have 2.0.27 standalone bridge installed already -> you run desktop -> it sees running bridge so it does not start built-in bridge. 
2. It might be fixable in connect, there is a workaround for older bridges but it has never worked properly.  https://github.com/trezor/trezor-suite/blob/3d16aee2e4ff94d808292297d5e4e8e513931147/packages/connect/src/device/DeviceCommands.ts#L690-L696

**Comment by mroz22** [comment]
Or maybe you could render the "x" button based on bridge version that is running.

File: packages/components/src/components/prompts/ConfirmOnDevice/index.tsx (hunk 2 above)
Comment (dahaca) [reply packages/components/src/components/prompts/ConfirmOnDevice/index.tsx]: Agree but I'd do an undefined check so TS doesn't complain that `activeStep` `Object is possibly 'undefined'.` in comparisons below.

File: packages/suite/src/components/suite/images/images.ts (hunk 3 above)
Comment (dahaca) [reply packages/suite/src/components/suite/images/images.ts:150]: Yaaay, thanks, I finally used a typeguard 😅

File: packages/suite/src/components/firmware/CheckSeed.tsx (hunk 4 above)
Comment (dahaca) [reply packages/suite/src/components/firmware/CheckSeed.tsx]: Paying my respects to the culture!

File: packages/suite/src/components/suite/images/DeviceConfirmImage.tsx (hunk 7 above)
Comment (dahaca) [reply packages/suite/src/components/suite/images/DeviceConfirmImage.tsx]: I see that `getDeviceModel` returns `'T' | '1'` but in [many places](https://github.com/trezor/trezor-suite/blob/d1d62b52efdb7aabedef0c30472e57c76765bbb4/packages/suite/src/components/firmware/ReconnectDevicePrompt.tsx#L232) `1 | 2` is expected. 🤯 I really think we need `enum TrezorModel` it's so unsettling for me to not have literally the _most important enum_

**Comment by dahaca** [comment]
@mroz22 @tsusanka Since it's related to Bridge ver, it makes sense to rely on it, not the platform. Please confirm that the only ver that's not supporting the `.cancel()` method is `'2.0.27'` :)

And should I be checking `state.suite.transport.bridge.version` or `state.suite.transport.version`?

**Comment by tsusanka** [comment]
> Please confirm that the only ver that's not supporting the .cancel() method is '2.0.27' :)

The other way around, v2.0.31 is the only one that supports it fully 🙂. I would simply make it >= v2.0.31 and that's it (future-ready when some new version gets released).

Also cc-ing @ondracja for context when this moves to QA.

**Comment by dahaca** [comment]
@tsusanka Ok! I am removing the **blocked** label then. Are there plans to make it work in earlier versions, should I create an issue?

**Comment by tsusanka** [comment]
> @tsusanka Ok! I am removing the blocked label then. Are there plans to make it work in earlier versions, should I create an issue?

Let's not worry about that, it is very minor part of our user base.

File: packages/suite/src/views/firmware/index.tsx (hunk 1 above)
Comment (dahaca) [reply packages/suite/src/views/firmware/index.tsx]: I didn't notice it affecting anything, removed 

cbd767025bd9279bc2d96c7ad5381c64bd6c3767

Comment (dahaca) [reply packages/suite/src/views/firmware/index.tsx]: Also no, it should not look like this 😭  Apparently it was long before this PR though...

cbd767025bd9279bc2d96c7ad5381c64bd6c3767

**Comment by dahaca** [comment]
@tsusanka @mroz22 So I see that the latest bundled version of bridge for desktop is 2.0.27 so the button is gone, although it works well. When will it be updated?

cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/suite/src/components/settings/SettingsMenu.tsx (hunk 9 above)
Comment (dahaca) [reply packages/suite/src/components/settings/SettingsMenu.tsx]: cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/components/src/components/prompts/ConfirmOnDevice/index.tsx (hunk 8 above)
Comment (dahaca) [reply packages/components/src/components/prompts/ConfirmOnDevice/index.tsx]: cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/suite/src/components/suite/images/DeviceConfirmImage.tsx (hunk 7 above)
Comment (dahaca) [reply packages/suite/src/components/suite/images/DeviceConfirmImage.tsx]: I volunteer to go thought he codebase and make all instances of major version and device model use the same `enum TrezorModel` :D For not switched to `getDeviceModel`

Also solved the touch problem 

cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/suite/src/components/suite/modals/confirm/ConfirmActionModal.tsx (hunk 6 above)
Comment (dahaca) [reply packages/suite/src/components/suite/modals/confirm/ConfirmActionModal.tsx]: cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/suite/src/components/suite/Modal/DevicePropmptModal.tsx (hunk 5 above)
Comment (dahaca) [reply packages/suite/src/components/suite/Modal/DevicePropmptModal.tsx]: cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/suite/src/components/firmware/CheckSeed.tsx (hunk 4 above)
Comment (dahaca) [reply packages/suite/src/components/firmware/CheckSeed.tsx]: cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/suite/src/components/suite/images/images.ts (hunk 3 above)
Comment (dahaca) [reply packages/suite/src/components/suite/images/images.ts:150]: cbd767025bd9279bc2d96c7ad5381c64bd6c3767

File: packages/components/src/components/prompts/ConfirmOnDevice/index.tsx (hunk 2 above)
Comment (dahaca) [reply packages/components/src/components/prompts/ConfirmOnDevice/index.tsx]: cbd767025bd9279bc2d96c7ad5381c64bd6c3767

**Comment by tsusanka** [comment]
Wait, Suite should have 2.0.31 bundled inside. Make sure you do not have a standalone bridge running before launching Suite (`pgrep trezord`).

**Review by tomasklim — APPROVED** [review]
LGTM

**Review by marekrjpolak — APPROVED** [review]

**Comment by bosomt** [comment]
edit: already solved
**QA OK**


**QA NOK**
I'm not sure if this is part of this modal change but i just noted strange artifact that's not on production version.
Please see video. 
When you open of detail  of signed transaction one element is present and not expected in tx detail section.

**Info:**
 - Suite version: desktop 22.6.0 (d72ff4c5b9f3cbe2ccb4be47170f88c08e9271d1)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/22.6.0 Chrome/94.0.4606.81 Electron/15.3.0 Safari/537.36
 - OS: MacIntel
 - Screen: 1680x1050
 - Device: model T 2.5.1 regular (revision a1d22890ee6a9ffec4edfc6e1b5b6c74d23855a3)
 - Transport: bridge 2.0.31


https://user-images.githubusercontent.com/31506317/170209180-6c6a1be4-6735-41df-9257-ef85a0f1dcfa.mov

**Comment by bosomt** [comment]
**QA OK**

**Info:**
 - Suite version: desktop 22.6.0 (2c0b29acd6d07eeb7ab417052791af1a670b2000)
 - Browser: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) TrezorSuite/22.6.0 Chrome/94.0.4606.81 Electron/15.3.0 Safari/537.36
 - OS: MacIntel
 - Screen: 1680x1050
 - Device: model T 2.4.3 regular (revision 595b14254c1abb2be3f69e42c7932f1eca8cf1b1)
 - Transport: bridge 2.0.27
 - Transport: bridge 2.0.31
